### PR TITLE
feat(1710): support external join. BREAKING CHANGE: add new external join logic [4]

### PR DESCRIPTION
**Pull request by d2lam**
Related: https://github.com/screwdriver-cd/screwdriver/issues/1710
Blocked by: https://github.com/screwdriver-cd/config-parser/pull/94

Previously, we only include nodes & edges from the workflow graph. This PR will generate external nodes & external edges that are downstream for each job.

### CODE REVIEW

File: lib/getWorkflow.js [hunk 1]
```diff
@@ -9,80 +9,203 @@
 const filterNodeName = name =>
     (/^~(pr|commit|release|tag|sd@)/.test(name) ? name : name.replace('~', ''));
```
Comment (tkyi) [inline lib/getWorkflow.js]:
```suggestion
    (/^~((pr|commit|release|tag)(?:$|:)|(sd@))/.test(name) ? name : name.replace('~', ''));
```

Might fix https://github.com/screwdriver-cd/screwdriver/issues/1417

File: lib/getWorkflow.js [hunk 2]
```diff
@@ -9,80 +9,203 @@
...
+            jobs[jobName].requires.forEach(n => nodes.add(filterNodeName(n)));
         }
-    });
+
+        // downstream nodes
+        const srcName = `sd@${pipelineId}:${jobName}`;
+
+        // Calculate which downstream jobs are triggered BY current job
+        // Only need to take care of external triggers, since internal will be taken care automatically
+        const externalDownstreamOr = await triggerFactory.getDestFromSrc(`~${srcName}`);
```
Comment (tkyi) [inline lib/getWorkflow.js]: where is the `getDestFromSrc` function defined?

File: lib/getWorkflow.js [hunk 3]
```diff
@@ -9,80 +9,203 @@
...
+        const externalDownstreamAnd = await triggerFactory.getDestFromSrc(srcName);
+
+        externalDownstreamOr.forEach((dest) => {
+            nodes.add(dest);
+        });
+
+        await buildExternalNodes(
+            jobName, externalDownstreamAnd, nodes, triggerFactory);
```
Comment (tkyi) [inline lib/getWorkflow.js]:
```suggestion
            srcName, externalDownstreamAnd, nodes, triggerFactory);
```

File: lib/getWorkflow.js [hunk 4]
```diff
@@ -9,80 +9,203 @@
...
+        if (Array.isArray(job.requires)) {
+            // Calculate which upstream jobs trigger the current job
+            const upstreamOr = new Set(
+                job.requires.filter(name => name.charAt(0) === '~'));
+            const upstreamAnd = new Set(
+                job.requires.filter(name => name.charAt(0) !== '~'));
+            const isJoin = upstreamAnd.size > 1;
+
+            upstreamOr.forEach((src) => {
+                edges.push({ src: filterNodeName(src), dest: jobName });
```
Comment (tkyi) [inline lib/getWorkflow.js]: do we need to call `filterNodeName` for dest/`jobName` as well?

File: lib/getWorkflow.js [hunk 5]
```diff
@@ -9,80 +9,203 @@
...
+            });
+            upstreamAnd.forEach((src) => {
+                const obj = { src, dest: jobName };
```
Comment (tkyi) [inline lib/getWorkflow.js]: do we need `filterNodeName` here?

File: lib/getWorkflow.js [hunk 6]
```diff
@@ -9,80 +9,203 @@
...
+
+        const externalDownstreamOr = await triggerFactory.getDestFromSrc(`~${srcName}`);
+        const externalDownstreamAnd = await triggerFactory.getDestFromSrc(srcName);
+
+        externalDownstreamOr.forEach((dest) => {
+            edges.push({ src: jobName, dest });
+        });
+
+        await buildExternalEdges(
+            jobName, externalDownstreamAnd, edges, triggerFactory);
```
Comment (tkyi) [inline lib/getWorkflow.js]:
```suggestion
            srcName, externalDownstreamAnd, edges, triggerFactory);
```

File: lib/getWorkflow.js (hunk 2 above)
Comment (d2lam) [reply lib/getWorkflow.js]: Good catch. It needs to be added to this PR: https://github.com/screwdriver-cd/models/pull/400/files

File: lib/getWorkflow.js (hunk 1 above)
Comment (d2lam) [reply lib/getWorkflow.js]: Can fix in a separate PR

Comment (tkyi) [reply lib/getWorkflow.js]: opened a PR: https://github.com/screwdriver-cd/workflow-parser/pull/28

File: lib/getWorkflow.js (hunk 2 above)
Comment (tkyi) [reply lib/getWorkflow.js]: added ✅

File: lib/getWorkflow.js (hunk 5 above)
Comment (tkyi) [reply lib/getWorkflow.js]: nvm don't need it because AND should not have `~`

File: lib/getWorkflow.js (hunk 4 above)
Comment (tkyi) [reply lib/getWorkflow.js]: nvm current job should not need it

File: lib/getWorkflow.js (hunk 6 above)
Comment (tkyi) [reply lib/getWorkflow.js]: actually looks like just `jobName` is fine here

File: test/lib/getWorkflow.test.js [hunk 7]
```diff
@@ -1,41 +1,48 @@
...
 const EXPECTED_EXTERNAL = require('../data/expected-external');
+const EXPECTED_EXTERNAL_COMPLEX = require('../data/expected-external-complex');
 
 NO_EDGES.edges = [];
 
-describe('getWorkflow', () => {
-    it('should throw if it is not given correct input', () => {
-        assert.throws(() => getWorkflow({ config: {} }),
-            Error, 'No Job config provided');
+describe.only('getWorkflow', () => {
```
Comment (DekusDenial) [inline test/lib/getWorkflow.test.js]: `.only`

Comment (tkyi) [reply test/lib/getWorkflow.test.js]: oops

**Pull request merged**